opensquirrel/merger/general_merger.py

```diff
@@ -79,6 +91,51 @@ def try_name_anonymous_bloch(bsr: BlochSphereRotation) -> BlochSphereRotation:
     return bsr
 
 
+def merge_barriers(statement_list: list[Statement]) -> list[Statement]:
+    """Function to fix the placement of the barriers such that the barriers are
+    merged in the circuit. Note that this function requires the barriers to be
+    placed in the correct order.
+
+    Args:
+        statement_list: list of statements
+
+    Returns:
+        Statement list with the barriers merged.
+    """
+    barrier_list: list[Barrier] = []
+    ordered_statement_list: list[Statement] = []
+    for _, statement in enumerate(statement_list):
+        if isinstance(statement, Gate):
+            if len(barrier_list) > 0:
+                barrier_qubits = next(barrier.get_qubit_operands() for barrier in barrier_list)[0]
+                if barrier_qubits in statement.get_qubit_operands():
+                    ordered_statement_list.extend(barrier_list)
+                    barrier_list = []
+            ordered_statement_list.append(statement)
+
+        if isinstance(statement, Barrier):
+            barrier_list.append(statement)
+
+    if len(barrier_list) > 0:
+        ordered_statement_list.extend(barrier_list)
+
+    return ordered_statement_list
+
+
+def optimise_circuit(circuit: Circuit) -> None:
+    """Optimise the circuit by merging the single qubit gates
+    and the barriers vertically.
+
+    Args:
+        circuit: Circuit object to optimise.
+
+    Returns:
+        None
+    """
+    merge_single_qubit_gates(circuit)
+    circuit.ir.statements = merge_barriers(circuit.ir.statements)
```

Comment:
I wouldn't probably add this function. I would just add the merge_barriers call at the end of the current merge_single_qubit_gates.